### PR TITLE
NIFI-1322 - PutHDFS - allow file append resolution

### PR DESCRIPTION
**Pull request by pvillard31**
Two remarks:
- LocalFileSystem object used for unit tests does not implement the append method.
- Based on HDFS-4600 (https://issues.apache.org/jira/browse/HDFS-4600), appending to a file is only possible if the replication factor is lesser or equal than the number of data nodes in the cluster. For live testing with a standalone instance, consider changing the replication factor to 1 for the file you are appending to.

Also, I considered that it would be up to the user to handle the demarcator between two appended messages if such a demarcator is expected.

Finally, if two processors are appending to the same file, errors may be raised due to concurrent access to the file. However, if redirecting the failure relationship back to the processor, the processors should, in the end, be able to append the messages. Obviously, in such a case, message ordering shall not be a requirement. Local testing suggests that users should use a MergeContent processor as upstream processor in order to ensure acceptable bandwidth.

### CODE REVIEW

File: nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/PutHDFS.java [hunk 1]
```diff
@@ -74,9 +75,15 @@
 @SeeAlso(GetHDFS.class)
 public class PutHDFS extends AbstractHadoopProcessor {
 
-    public static final String REPLACE_RESOLUTION = "replace";
-    public static final String IGNORE_RESOLUTION = "ignore";
-    public static final String FAIL_RESOLUTION = "fail";
```
Comment (olegz) [inline nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/PutHDFS.java]: Pierre; I understand why/what you did here, but given the fact that both the class and the variable names are ```public```, this constitutes a breaking change.

File: nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/PutHDFS.java [hunk 2]
```diff
@@ -315,21 +328,24 @@ public void process(InputStream in) throws IOException {
...
-                    break;// rename was successful
+            if(!conflictResponse.equals(APPEND_RESOLUTION.getValue())
+                    || (conflictResponse.equals(APPEND_RESOLUTION.getValue()) && !destinationExists)) {
+                boolean renamed = false;
+                for (int i = 0; i < 10; i++) { // try to rename multiple times.
+                    if (hdfs.rename(tempCopyFile, copyFile)) {
+                        renamed = true;
+                        break;// rename was successful
+                    }
+                    Thread.sleep(200L);// try waiting to let whatever might cause rename failure to resolve
```
Comment (olegz) [inline nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/PutHDFS.java]: Do you know what's this for and if we still need it? I mean not only the sleep, but he whole loop.

File: nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/PutHDFS.java (hunk 1 above)
Comment (pvillard31) [inline nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/PutHDFS.java]: Oh yes, you're right! Will update this part asap!

File: nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/PutHDFS.java (hunk 2 above)
Comment (pvillard31) [inline nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/PutHDFS.java]: To be honest, no I don't. I assume that there is some kind of asynchronous call we are doing before when talking to remote clusters and that we might need to make this try multiple times to actually succeed. Not sure if it is still required with the last versions of HDFS client but didn't want to change it.

Comment (olegz) [inline nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/PutHDFS.java]: Fair enough, but probably needs to be revisited at some point in the future.

**Pull request closed without merging**